partition.go

```diff
@@ -0,0 +1,141 @@
+package schema
+
+import (
+	"encoding/binary"
+	"hash/fnv"
+	"sort"
+
+	"github.com/cespare/xxhash"
+	jump "github.com/dgryski/go-jump"
+)
+
+type PartitionByMethod uint8
+
+const (
+	PartitionByOrg PartitionByMethod = iota
+	PartitionBySeries
+	PartitionBySeriesWithTags
+	PartitionBySeriesWithTagsFnv
+)
+
+func (m *MetricData) PartitionID(method PartitionByMethod, partitions int32) (int32, error) {
+	var partition int32
+
+	switch method {
+	case PartitionByOrg:
+		h := fnv.New32a()
+		err := binary.Write(h, binary.LittleEndian, uint32(m.OrgId))
+		if err != nil {
+			return 0, err
+		}
+		partition = int32(h.Sum32()) % partitions
+		if partition < 0 {
+			partition = -partition
+		}
+	case PartitionBySeries:
+		h := fnv.New32a()
+		h.Write([]byte(m.Name))
+		partition = int32(h.Sum32()) % partitions
+		if partition < 0 {
+			partition = -partition
+		}
+	case PartitionBySeriesWithTags:
+		h := xxhash.New()
+		h.WriteString(m.Name)
+		sort.Strings(m.Tags)
```

Comment:
this procedure where the tags get sorted and then concatenated into a string (excluding `name`) is repeated multiple times, that could also just be in a private method like `.getSortedTagString()`

Reply:
> Should we perform a check to see if any tags exist?

we probably should not spend any effort on this or on seeing whether we should, as this is unlikely to be an issue

> Even if tags don't exist I suppose we would still want to continue with the jump.Hash call on m.Name

yes, in that case it's equivalent to PartitionBySeries, which is what we want

> this procedure where the tags get sorted and then concatenated into a string (excluding name) is repeated multiple times, that could also just be in a function like getSortedTagString([]string) string

+1

Reply:
>  could also just be in a function like getSortedTagString([]string) string

We are deliberately optimising here to reduce allocations.  We dont want to allocate memory for a new string only to then write that string to something else.

But we could use a func like `writeSortedTagString(w io.StringWriter, []string) (n int, err error)`

Reply:
actually, that wont work, as we are using WriteString() when it is available and Write() otherwise. 
so we would need to just use `writeSortedTagString(w io.Writer, []string) (n int, err error)`   in which we can call `io.WriteString(w, s)` which will use the optimised WriteString() call if available and fall back to Write()

Reply:
I implemented that and pushed into this branch